File: shared/src/engine/meld-validator.ts
import { Card, Rank, Suit } from '../types/card';
import { Meld, MeldType } from '../types/meld';
import { GameConfig } from '../types/game-config';

/** Check if cards form a valid set (3-4 same rank, different suits) */
export function isValidSet(cards: Card[], config: GameConfig): boolean {
  if (cards.length < 3 || cards.length > 4) return false;

  const jokers = cards.filter(c => c.isJoker);
  const normals = cards.filter(c => !c.isJoker);

  if (jokers.length > config.maxJokersPerMeld) return false;
  if (normals.length === 0) return false;

  // All normal cards must have the same rank
  const rank = normals[0].rank!;
  if (!normals.every(c => c.rank === rank)) return false;

  // All suits must be different (among normal cards)
  const suits = new Set(normals.map(c => c.suit!));
  if (suits.size !== normals.length) return false;

  return true;
}

/** Check if cards form a valid run (3+ consecutive, same suit) */
export function isValidRun(cards: Card[], config: GameConfig): boolean {
  if (cards.length < 3) return false;

  const jokers = cards.filter(c => c.isJoker);
  const normals = cards.filter(c => !c.isJoker);

  if (jokers.length > config.maxJokersPerMeld) return false;
  if (normals.length === 0) return false;

  // All normal cards must have the same suit
  const suit = normals[0].suit!;
  if (!normals.every(c => c.suit === suit)) return false;

  // Try Ace as low (rank 1) first
  if (tryRunWithAceValue(normals, jokers.length, false)) return true;

  // If any Ace present, retry with Ace as high (rank 14)
  const hasAce = normals.some(c => c.rank === Rank.Ace);
  if (hasAce && tryRunWithAceValue(normals, jokers.length, true)) return true;

  return false;
}

function tryRunWithAceValue(normals: Card[], jokerCount: number, aceHigh: boolean): boolean {
  const getRank = (c: Card) => aceHigh && c.rank === Rank.Ace ? 14 : c.rank!;

  const sorted = [...normals].sort((a, b) => getRank(a) - getRank(b));

  // Check for duplicates
  for (let i = 1; i < sorted.length; i++) {
    if (getRank(sorted[i]) === getRank(sorted[i - 1])) return false;
  }

  // Check consecutive with jokers filling gaps
  return canFormConsecutiveRun(sorted.map(c => getRank(c)), jokerCount);
}

function canFormConsecutiveRun(sortedRanks: number[], jokerCount: number): boolean {
  if (sortedRanks.length === 0) return jokerCount >= 3;

  let jokersUsed = 0;
  let prev = sortedRanks[0];

  for (let i = 1; i < sortedRanks.length; i++) {
    const curr = sortedRanks[i];
    const gap = curr - prev - 1;
    if (gap < 0) return false; // duplicate
    jokersUsed += gap;
    if (jokersUsed > jokerCount) return false;
    prev = curr;
  }

  return true;
}

/** Validate a meld (auto-detect type) */
export function validateMeld(cards: Card[], config: GameConfig): MeldType | null {
  if (isValidSet(cards, config)) return MeldType.Set;
  if (isValidRun(cards, config)) return MeldType.Run;
  return null;
}

/** Calculate the point value of a meld (for opening threshold) */
export function calculateMeldPoints(meld: Meld, config: GameConfig): number {
  if (meld.type === MeldType.Run) {
    return calculateRunMeldPoints(meld, config);
  }
  let total = 0;
  for (const card of meld.cards) {
    total += getCardPoints(card, config);
  }
  return total;
}

/**
 * Calculate run meld points with context-aware Ace value:
 * A-2-3 → Ace = 1 (low); Q-K-A → Ace = 11 (high)
 * Jokers take the point value of the position they fill.
 */
function calculateRunMeldPoints(meld: Meld, config: GameConfig): number {
  const normals = meld.cards.filter(c => !c.isJoker);
  const jokerCount = meld.cards.length - normals.length;

  if (normals.length === 0) {
    return jokerCount * config.jokerValue;
  }

  const hasAce = normals.some(c => c.rank === Rank.Ace);

  // Determine if Ace is high: if any non-ace normal card has rank >= Queen
  let aceIsHigh = false;
  if (hasAce) {
    const nonAceRanks = normals.filter(c => c.rank !== Rank.Ace).map(c => c.rank!);
    aceIsHigh = nonAceRanks.length > 0 && Math.max(...nonAceRanks) >= Rank.Queen;
  }

  const getRank = (c: Card) => aceIsHigh && c.rank === Rank.Ace ? 14 : c.rank!;

  // Build sorted normal rank values
  const sortedRanks = normals.map(c => getRank(c)).sort((a, b) => a - b);

  // Build full sequence of rank positions (fill joker gaps)
  const minRank = sortedRanks[0];
  const positions: number[] = [];
  let idx = 0;
  let r = minRank;
  while (positions.length < meld.cards.length) {
    if (idx < sortedRanks.length && sortedRanks[idx] === r) {
      positions.push(r);
      idx++;
    } else {
      positions.push(r); // joker fills this position
    }
    r++;
  }

  // Sum values for each position
  const rankPointValue = (rank: number): number => {
    if (rank === 14) return config.aceHighValue; // Ace high
    if (rank >= Rank.Jack) return 10; // J/Q/K
    return rank; // 1-10
  };

  return positions.reduce((sum, pos) => sum + rankPointValue(pos), 0);
}

/** Get point value of a single card */
export function getCardPoints(card: Card, config: GameConfig): number {
  if (card.isJoker) return config.jokerValue;
  const rank = card.rank!;
  if (rank === Rank.Ace) return config.aceHighValue;
  if (rank >= Rank.Jack) return 10;
  return rank;
}

/** Build a Meld object from validated cards */
export function buildMeld(
  id: string,
  cards: Card[],
  type: MeldType,
): Meld {
  const jokerSubs: Record<number, { suit: string; rank: number }> = {};
  // For now, joker substitutions are computed when needed
  return { id, type, cards, jokerSubstitutions: jokerSubs };
}

/** Check if a card can be appended to an existing meld */
export function canLayoff(
  card: Card,
  meld: Meld,
  position: 'start' | 'end',
  config: GameConfig,
): boolean {
  const newCards = position === 'start'
    ? [card, ...meld.cards]
    : [...meld.cards, card];

  if (meld.type === MeldType.Set) return isValidSet(newCards, config);
  if (meld.type === MeldType.Run) return isValidRun(newCards, config);
  return false;
}

File: mobile/lib/engine/meld_validator.dart
import '../models/card.dart';
import '../models/meld.dart';
import '../models/game_config.dart';

/// Check if cards form a valid set
bool isValidSet(List<Card> cards, GameConfig config) {
  if (cards.length < 3 || cards.length > 4) return false;

  final jokers = cards.where((c) => c.isJoker).toList();
  final normals = cards.where((c) => !c.isJoker).toList();

  // Max jokers per meld check (matching server)
  if (jokers.length > config.maxJokersPerMeld) return false;

  // Need at least 1 real card to determine rank
  if (normals.isEmpty) return false;

  final rank = normals.first.rank!;
  if (!normals.every((c) => c.rank == rank)) return false;

  final suits = normals.map((c) => c.suit).toSet();
  if (suits.length != normals.length) return false;

  return true;
}

/// Check if cards form a valid run
bool isValidRun(List<Card> cards, GameConfig config) {
  if (cards.length < 3) return false;

  final jokers = cards.where((c) => c.isJoker).toList();
  final normals = cards.where((c) => !c.isJoker).toList();

  // Max jokers per meld check (matching server)
  if (jokers.length > config.maxJokersPerMeld) return false;

  // Need at least 1 real card to determine suit/rank anchor
  if (normals.isEmpty) return false;

  final suit = normals.first.suit;
  if (!normals.every((c) => c.suit == suit)) return false;

  // Try Ace as low (1) first, then as high (14)
  final hasAce = normals.any((c) => c.rank!.value == 1);

  bool tryConsecutive(int aceValue) {
    final values = normals.map((c) => c.rank!.value == 1 ? aceValue : c.rank!.value).toList()..sort();
    // Check duplicates
    for (int i = 1; i < values.length; i++) {
      if (values[i] == values[i - 1]) return false;
    }
    // Check consecutive with jokers
    int used = 0;
    for (int i = 1; i < values.length; i++) {
      final gap = values[i] - values[i - 1] - 1;
      if (gap < 0) return false;
      used += gap;
      if (used > jokers.length) return false;
    }
    return true;
  }

  if (tryConsecutive(1)) return true;
  if (hasAce && tryConsecutive(14)) return true;
  return false;
}

/// Validate a meld (auto-detect type)
MeldType? validateMeld(List<Card> cards, GameConfig config) {
  if (isValidSet(cards, config)) return MeldType.set;
  if (isValidRun(cards, config)) return MeldType.run;
  return null;
}

/// Get point value of a card (standalone — used for PENALTY scoring at end of round)
/// Joker = 30 points penalty. As = 11 penalty.
int getCardPoints(Card card, GameConfig config) {
  if (card.isJoker) return config.jokerValue; // 30
  final rank = card.rank!.value;
  if (rank == 1) return config.aceHighValue; // 11
  if (rank >= 11) return 10;
  return rank;
}

/// Get the point value of a single rank for meld scoring (not penalty).
/// Ace low (value=1) = 1 point. Ace high (value=14) = 11 points.
int _rankPoints(int rankValue, GameConfig config) {
  if (rankValue == 14) return config.aceHighValue; // Ace high = 11
  if (rankValue == 1) return 1; // Ace low = 1
  if (rankValue >= 11) return 10; // J, Q, K = 10
  return rankValue;
}

/// Calculate meld points for opening — Joker takes the value of
/// the card it replaces (e.g. 10♠ 9♠ Joker → Joker = 8♠ = 8 pts).
/// Ace = 1 if low (A-2-3), 11 if high (Q-K-A).
int calculateMeldPoints(List<Card> cards, GameConfig config) {
  final type = validateMeld(cards, config);
  if (type == null) {
    // Fallback: sum using penalty values
    return cards.fold(0, (sum, c) => sum + getCardPoints(c, config));
  }

  final jokers = cards.where((c) => c.isJoker).toList();
  final normals = cards.where((c) => !c.isJoker).toList();

  if (type == MeldType.set) {
    // Set: all same rank. Joker replaces that same rank.
    // Joker takes the value of the rank it replaces, not 30.
    final realCards = cards.where((c) => !c.isJoker).toList();
    final rankValue = realCards.isNotEmpty ? getCardPoints(realCards.first, config) : config.jokerValue;
    return cards.length * rankValue;
  }

  // Run: consecutive in same suit. Jokers fill gaps.
  // Determine ace interpretation (low=1 or high=14)
  final hasAce = normals.any((c) => c.rank!.value == 1);

  // To decide ace value: check which interpretation makes a valid run
  int aceValue = 1; // default low
  if (hasAce) {
    // Try as high (14) — check if other cards are in high range (10+)
    final otherValues = normals.where((c) => c.rank!.value != 1).map((c) => c.rank!.value).toList();
    if (otherValues.isNotEmpty) {
      final maxOther = otherValues.reduce((a, b) => a > b ? a : b);
      // If highest non-ace card is >= 10 (J/Q/K), ace is high
      if (maxOther >= 10) aceValue = 14;
    }
  }

  final normalValues = normals
      .map((c) => c.rank!.value == 1 ? aceValue : c.rank!.value)
      .toList()..sort();

  // Build complete sequence filling in joker positions
  final fullSequence = <int>[];
  int jokerIdx = 0;
  if (normalValues.isEmpty) {
    return cards.fold(0, (sum, c) => sum + getCardPoints(c, config));
  }

  int current = normalValues.first;
  int normalPtr = 0;
  while (fullSequence.length < cards.length) {
    if (normalPtr < normalValues.length && normalValues[normalPtr] == current) {
      fullSequence.add(current);
      normalPtr++;
    } else if (jokerIdx < jokers.length) {
      fullSequence.add(current); // Joker takes this value
      jokerIdx++;
    } else {
      break; // Safety
    }
    current++;
  }

  // Sum all values using _rankPoints
  return fullSequence.fold(0, (sum, v) => sum + _rankPoints(v, config));
}

/// Check if a card can be laid off onto an existing meld
bool canLayoff(Card card, Meld meld, String position, GameConfig config) {
  final newCards = position == 'start'
      ? [card, ...meld.cards]
      : [...meld.cards, card];

  if (meld.type == MeldType.set) return isValidSet(newCards, config);
  if (meld.type == MeldType.run) return isValidRun(newCards, config);
  return false;
}

/// Check if a card can be laid off at either end, or swap a joker
bool canLayoffAny(Card card, Meld meld, GameConfig config) {
  if (canLayoff(card, meld, 'start', config)) return true;
  if (canLayoff(card, meld, 'end', config)) return true;

  // Can it replace a joker in this meld?
  final jokers = meld.cards.where((c) => c.isJoker).toList();
  if (jokers.isNotEmpty && !card.isJoker) {
    for (final joker in jokers) {
      final idx = meld.cards.indexOf(joker);
      final testCards = List<Card>.from(meld.cards);
      testCards[idx] = card;
      if (validateMeld(testCards, config) != null) return true;
    }
  }
  return false;
}

/// A "clean run" is a Run (suite) that contains zero jokers.
bool isCleanRun(Meld meld) {
  return meld.type == MeldType.run && meld.cards.every((c) => !c.isJoker);
}

/// Check if the opening melds meet the opening requirements:
/// - Total points >= openingThreshold
/// - At least one clean run if openingRequiresCleanRun
({bool valid, int points, String? reason}) validateOpening(
  List<Meld> melds,
  GameConfig config,
) {
  final points = melds.fold<int>(
    0, (sum, m) => sum + calculateMeldPoints(m.cards, config),
  );

  if (config.openingThreshold > 0 && points < config.openingThreshold) {
    return (
      valid: false,
      points: points,
      reason: 'Il faut ${config.openingThreshold} points pour ouvrir (vous avez $points)',
    );
  }

  if (config.openingRequiresCleanRun && !melds.any((m) => isCleanRun(m))) {
    return (
      valid: false,
      points: points,
      reason: 'L\'ouverture nécessite au moins une suite sans joker',
    );
  }

  return (valid: true, points: points, reason: null);
}






